### ISSUE TITLE
mbedtls_x509_crt_parse_path() returns success when no certificates are loaded.

### ISSUE THREAD

**Issue by jp-bennett** [ARMmbed/mbedtls#3009]
The Linux implementation of mbedtls_x509_crt_parse_path() doesn't seem to check that valid certificates were loaded. If that function doesn't attempt to parse any files, it returns 0. So, call parse_path on an empty folder, and you get a success result. That doesn't seem to be intended. The Windows implementation, for instance, returns an IO error if the given path is empty. It seems like there should be a final check just before returning, that at least one valid certificate was loaded onto the chain.

**Comment by yanesca** [comment]
Thank you for opening this issue!

Just a remark about the suggested solution: I think that to align the Linux implementation's behaviour with the function documentation and the behaviour on Windows, it is probably better to return an error when the given path is empty (instead of making a check in the end).

**Comment by jp-bennett** [comment]
I discovered this issue while troubleshooting the symlink issue. In that case, the target directory isn't empty, but none of the objects are processed.

This was an unexpected corner case, and it made the bug I was tracking harder to pin down. I suggested a check at the end to avoid playing wack-a-mole with other corner cases. Example, when the given directory doesn't contain any files, but contains other directories. I believe the current code would return success in that case as well, maybe even in the win32 code. 

The documentation seems to say that success should only be returned if a certificate file parses successfully. This seems to imply to me that we should have at least one cert in the chain.

Another option, define a variable at the top of the function, have_processed_cert = 0. When a file loads successfully, set it to 1. At the end of the function, check that it's set to 1, else return an error.

**Comment by yanesca** [comment]
The documentation is slightly ambiguous about the return value when no certificates have been loaded, but the function did not encounter any errors. I agree that it makes sense to handle these cases in a uniform manner.

You are making a good point that semantically it makes more sense to return an error in this case, I was trying to adjust the solution to my reading of the documentation. But you are making a good argument for rather going for what makes the most sense. We just need to update/clarify the documentation as well.

**Comment by jp-bennett** [comment]
> The documentation is slightly ambiguous about the return value when no certificates have been loaded, but the function did not encounter any errors.
True. I can imagine a scenario where I would want to load all the certs from a directory, and if there's nothing there, that's fine. In the OBS scenario, we want to load the certs from their default location, and if no certs got loaded, our hair is on fire.

It seems like the project devs need to just make a call one way or the other, whether this function errors out in the case of no files being processed, and then clarify the documentation to reflect that decision. Currently, the docs assume that there will always be at least one file in the given path, so the outcome will always be either a failure code, or a partial success.

Once it's settled how the function should behave, it will be trivial to make the 3 different implementations behave the same.

**Comment by gilles-peskine-arm** [comment]
@jp-bennett writes:

> If that function doesn't attempt to parse any files, it returns 0. So, call parse_path on an empty folder, and you get a success result.

That's the way I understand the documentation. Quoting it:

> `\return` 0 if all certificates parsed successfully, a positive number if partly successful or a specific X509 or PEM error code

This is programming documentation, so I expect “all” to have the logician's meaning (every certificate that it encountered is parsed successfully), and not the common language meaning where this implicitly claims that at least one certificate was parsed successfully.

The documentation also does state

> Load one or more certificate files

which rules out loading zero certificate files. The thing is, if no certificate file is found, there's no error to return. So the only permitted return value is 0, even if the first sentence in the description implies that this isn't an intended outcome.

I don't understand why it wouldn't be desirable for `mbedtls_x509_crt_parse_path` to indicate success when no certificates are found. Could you explain why you think there should be an exception @jp-bennett ?

To me this behavior seems very natural. It's common for applications to have multiple directories for configuration files, some of which may or may not have files in them.
```
char *paths[] = {
    "/etc/ca-certificates",
    "/usr/share/ca-certificates",
    "/usr/local/share/ca-certificates",
    strcat(getenv("HOME"), "/.config/ca-certificates")), // not correct C code but you know what I mean
};
for (char **pp = paths; *pp; ++pp) {
    if (access(*pp, F_OK)) {
        ret = mbedtls_x509_crt_parse_path(chain, *pp);
        if (ret < 0) goto fail;
    }
}
```
If one of these directories exists but contains no certificates, how can I tell that this is the problem and the loop should continue?

I do agree that the behavior should be consistent between platforms. Regardless of which behavior we select, we need to provide a way to adapt existing code, so we should be more precise in the explanation of error codes.

I lean towards allowing empty directories on all platforms, both because I think it's the right behavior (it's more uniform) and because I think it's less likely to break existing applications (what application would rely on getting an error from empty directories?).

**Comment by jp-bennett** [comment]
> I lean towards allowing empty directories on all platforms, both because I think it's the right behavior (it's more uniform) and because I think it's less likely to break existing applications (what application would rely on getting an error from empty directories?).

I suspect many applications use the logic: If the return value is 0 or positive, then we must have a valid certificate chain.

If an application is checking multiple directories, it should already be prepared to handle one of those directories returning an error. An error value indicating "no certificates loaded" seems like a reasonable error.

I don't think it greatly matters which behavior is chosen, some program will break as a result (https://xkcd.com/1172/). I don't have an especially strong opinion, I'd just like to see the behavior be consistent next time I have to troubleshoot an issue.

**Comment by gilles-peskine-arm** [comment]
> If the return value is 0 or positive, then we must have a valid certificate chain.

Ok, good point. The Unix behavior only guarantees that there's at least one certificate if the directory contains at least one certificate. Would this error be typically detected later when the application attempts to use the certificates, though, or would it typically be ignored, leading to bad outcomes?

We need to think carefully about backward compatibility for the sake of long-time support branches. We try to avoid backporting changes that break plausible applications, even spacebar heater applications, but here, given that the behavior is inconsistent between platforms, we really should fix all supported branches.

**Comment by rcr** [comment]
Any update on this one? I was under the impression from reading the documentation for this function that a return value of 0 implies at least one cert was successfully loaded. To support multiple platforms and distros, I want to check a few paths at startup, e.g.:


    /etc/ssl/certs
    /etc/ssl/


And stop when certs are successfully loaded. I can check for `MBEDTLS_ERR_X509_FILE_IO_ERROR` for non existent paths, so it would make sense to me to have another error to indicate that the directory exists, but no certs were found.
